src/python_framework_v02/troute/nhd_network_augment.py

```diff
@@ -159,49 +167,138 @@ def build_reaches(rconn):
     for tw, net in subnets.items():
         path_func = partial(nhd_network.split_at_junction, net)
         subreaches[tw] = nhd_network.dfs_decomposition(net, path_func)
+    
+    return subreachable, subreaches
 
-    return subreachable, subreaches, subnets
+def find_headwater_reaches(subreaches, hws):
+    
+    """
+    Construct a list of headwater reaches
+    Args:
+        subreaches (dict): reaches (linearly connected segments) connected to each network tailwater (keys)
+        hws (list): headwater segments
+    Returns:
+        hw_reaches (list): lists of headwater reaches 
+    """
+    
+    hw_reaches = []
+    for sublists in list(subreaches.values()):
+        for rch in sublists:
+            for val in rch:
+                if val in hws:
+                    hw_reaches.append(rch)
+                    pass
+    
+    return hw_reaches
+
+def find_short_headwater_reaches(hw_reaches, idx_array, length_array, threshold):
 
+    """
+    Construct a list of short headwater reaches
+    Args:
+        hw_reaches (list): lists of headwater reaches 
+        idx_array (numpy array): ordered network segment indices (link IDs)
+        length_array (numpy array): network segment lengths, ordered by idx_array
+        threshold (int): threshold length that defines short
+    Returns:
+        short_hw_reaches (list): lists of SHORT headwater reaches 
+    """
+    
+    short_hw_reaches = []
+    for rch in hw_reaches:
+
+        # find row locations of reach segments
+        a = np.searchsorted(idx_array, rch)
```

Comment:
Do we have a guarantee that idx_array is sorted? If idx_array not sorted, np.searchsorted will fail.

Reply:
Yes. `idx_array` is extracted from the network data DataFrame constructed and returned by `def get_network_data`.  In this function, the DataFrame is sorted by index - ensuring that `idx_array` will also be sorted.